E2E tests: Onboarding Step 2 - Configure product listings

ianlin opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes part of #2070, this PR adds E2E tests for [📌 Onboarding Step 2 - Configure product listings][1].

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Follow these steps to set up the e2e env: https://github.com/woocommerce/google-listings-and-ads#e2e-testing
2. Run `npm run test:e2e -- ./tests/e2e/specs/setup-mc/step-2-product-listings.test.js`, the test should pass.
3. Or, run `npm run test:e2e-dev -- ./tests/e2e/specs/setup-mc/step-2-product-listings.test.js`, check the test in the chromium step by step.
4. Run `npm run test:e2e` to make sure all tests are passed.

### Changelog entry

> Dev - E2E - Onboarding Step 2 - Configure product listings

[1]: https://github.com/woocommerce/google-listings-and-ads/issues/2070#onboarding-step-2

ianlin committed:
Add more mocking requests

ianlin committed:
Add ProductListingsPage for e2e test

ianlin committed:
Add e2e tests for onboarding step 2 - configure product listings

jorgemd24 commented on tests/e2e/utils/pages/setup-mc/step-2-product-listings.js:
```diff
@@ -0,0 +1,514 @@
+		return this.page.locator(
+			'.woocommerce-tree-select-control > .components-base-control'
+		);
+	}
```
I believe this part could be used in other tests because the tree selector is also used on different pages. In my next PR, I will try to put it in a generic file.

jorgemd24 commented on tests/e2e/utils/pages/setup-mc/step-2-product-listings.js:
```diff
@@ -0,0 +1,514 @@
+	 *
+	 * @param {string} name
+	 *
+	 * @return {import('@playwright/test').Locator} Get tree item by country name.
```
This should be `Get Remove country button`

jorgemd24 commented on tests/e2e/utils/pages/setup-mc/step-2-product-listings.js:
```diff
@@ -0,0 +1,514 @@
+	async goto() {
+		await this.page.goto(
+			'/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc&google-mc=connected',
+			{ waitUntil: LOAD_STATE.NETWORK_IDLE }
```
Do you think that we need NETWORK_IDLE? The docs say that NETWORK_IDLE is discouraged.

```
     * - `'networkidle'` - **DISCOURAGED** consider operation to be finished when there are no network connections for
     *   at least `500` ms. Don't use this method for testing, rely on web assertions to assess readiness instead.
```


 I changed to DOM_CONTENT_LOADED and still passed all tests.

jorgemd24 commented on tests/e2e/specs/setup-mc/step-2-product-listings.test.js:
```diff
@@ -0,0 +1,275 @@
+		await expect( countrySearchBoxContainer ).not.toContainText(
+			'United Kingdom (UK)'
+		);
+	} );
```
Shouldn't we test the default country (US in this case) should be set?

jorgemd24 commented on tests/e2e/specs/setup-mc/step-2-product-listings.test.js:
```diff
@@ -0,0 +1,275 @@
+test.describe.configure( { mode: 'serial' } );
+
+/**
+ * @type {import('../../utils/pages/setup-mc/step-1-set-up-accounts.js').default} productListingsPage
```
This definition is incorrect, it should be `step-2-product-listings.js`

jorgemd24 commented on tests/e2e/utils/pages/setup-mc/step-2-product-listings.js:
```diff
@@ -0,0 +1,514 @@
+			name: 'No',
+			exact: true,
+		} );
+	}
```
This could be one function that accepts one parameter, for example:

`getOfferFreeShippingForOrderRadioRow (name = 'yes')`

jorgemd24 commented on tests/e2e/specs/setup-mc/step-2-product-listings.test.js:
```diff
@@ -0,0 +1,275 @@
+
+	test.describe( 'Shipping rate is simple', () => {
+		test.beforeAll( async () => {
+			await productListingsPage.goto();
```
There is also the option to reload the page, `page.reload`, although I think it will have the same effect.

jorgemd24 reviewed (commented):
Thanks, @ianlin, for your work on this! 

The PR covers many scenarios, which is great 👍. The only thing I feel is missing is that when the "Continue" Button is clicked, it should take you to the next page so we can test the next page containing "Confirm store requirements", WDYT? Additionally, I've left some other comments related to the code.

ianlin committed:
Correct the JSDoc description for getRemoveCountryButtonByName function

ianlin committed:
Replace NETWORK_IDLE with DOM_CONTENT_LOADED in e2e tests

ianlin committed:
Add a test to check default country US in onboarding step 2 e2e test

ianlin committed:
Correct the type definition in onboarding step 2 e2e test

ianlin committed:
Combine two util functions to one in onboarding step 2 e2e test

ianlin committed:
Move page.goto to beforeAll and replace some page.goto to page.reload in
e2e tests

ianlin committed:
Add a test of "Click Continue button should see heading of next step"

ianlin committed:
Fix JSLint error

ianlin committed:
Add tests for checking whether requests are triggered in certain actions
in onboarding step 2 e2e tests

ianlin commented:
Thanks @jorgemd24 for reviewing and the nice suggestions. I've pushed more commits to address them and it's ready for the second round.

jorgemd24 reviewed (commented):
Thanks @ianlin for all the adjustments. Looks good but there's an issue with the last test - it's failing because the tax rate hasn't been selected.

![2hciJV.png](https://github.com/woocommerce/google-listings-and-ads/assets/2488994/76a1beaa-8b43-4496-a147-c4b10c21d1da)

ianlin committed:
Check tax rate radio before clicking continue button for onboarding step
2 e2e test

ianlin commented:
Thanks for catching it @jorgemd24! I found in my local it passed because I have selected a tax rate so it had been stored in the test DB. I added a commit to mock the `GET /gla/mc/settings` request and also updated the test to select the tax rate before clicking continue button in the last test. It's ready for another round of review. 🙏

jorgemd24 reviewed (approved):
Thanks @ianlin for the changes, LGTM 👍  All tests pass now.
